### PR TITLE
Adding Intent and Slot documentation

### PR DESCRIPTION
**Pull request by vladgets**
Signed-off-by: Vlad Getselevich <vgetselevich@nvidia.com>

### CODE REVIEW

File: docs/source/nlp/joint_intent_slot.rst [hunk 1]
```diff
@@ -3,10 +3,253 @@
...
+- :code:`dict.slots.csv` - A list of all Slot names in the data. Only one line per intent name
+  is allowed.
+- :code:`train.tsv/test.tsv` - A list of original queries, one per line, with the intent number
+  separated by a tab (e.g. "what alarms do i have set right now <TAB> 0"). Intent numbers are
+  set according to the intent line in the intent dictionary file (:code:`dict.intents.csv`),
+  starting from 0. The first line in these files should contain the header line "sentence
+  <tab> label".
+- :code:`train_slot.tvs/test_slot.tsv` - A list that contains one line per a query, when an each word from the original text queries,
+  is replaced by a token number from the slots dictionary file (dict.slots.csv), counted starting from 0.
+  All the words which do not contain a relevant slot are replaced by 'out-of scope' token number, which is also a part of the slot dictionary file,
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:45]: Could you provide an example of dicts files? this part `'out-of scope' token number, which is also a part of the slot dictionary file,` is unclear without an example

File: docs/source/nlp/joint_intent_slot.rst [hunk 2]
```diff
@@ -3,10 +3,253 @@
...
+    We recommend you try this model in a Jupyter notebook \
+    (can run on `Google's Colab <https://colab.research.google.com/notebooks/intro.ipynb>`_.): \
+    `NeMo/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb <https://github.com/NVIDIA/NeMo/blob/main/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb>`__.
+
+    Connect to an instance with a GPU (Runtime -> Change runtime type -> select "GPU" for hardware accelerator)
+
+    An example script on how to train the model could be found here: `NeMo/examples/nlp/intent_slot_classification <https://github.com/NVIDIA/NeMo/tree/main/examples/nlp/intent_slot_classification>`__.
+
+
+Data Format
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:28]: Is this `NeMo Format`? If so, please rename to `NeMo Data Format`

File: docs/source/nlp/joint_intent_slot.rst [hunk 3]
```diff
@@ -3,10 +3,253 @@
 Joint Intent and Slot Classification
 ====================================
 
-We recommend you try this model in a Jupyter notebook \
-(can run on `Google's Colab <https://colab.research.google.com/notebooks/intro.ipynb>`_.): \
-`NeMo/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb <https://github.com/NVIDIA/NeMo/blob/main/tutorials/nlp/Joint_Intent_and_Slot_Classification.ipynb>`__.
+Joint Intent and Slot classification is a method for classifying an Intent and detecting all
+relevant Slots (Entities) for the Intent in a query.
+For example, in the query `What is the weather in Santa Clara tomorrow morning?`,
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:8]: Maybe mention NLU task here?

File: docs/source/nlp/joint_intent_slot.rst [hunk 4]
```diff
@@ -3,10 +3,253 @@
...
+
+
+
+Model Training
+--------------
+
+This is a pretrained Bert based model with 2 linear classifier heads on the top of it,
+one for classifying an intent of the query and another for classifying slots for each token of the query.
+This model is trained with the combined loss function on the Intent and Slot classification task on the given dataset.
+The model architecture is based on the paper: https://arxiv.org/pdf/1902.10909.pdf
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:100]: Add link to the paper + reference to the paper

File: docs/source/nlp/joint_intent_slot.rst [hunk 5]
```diff
@@ -3,10 +3,253 @@
...
+
+For each query the model will classify it as one the intents from the intent dictionary and
+for each word of the query it will classify it as one of the slots from the slot dictionary, including out of scope slot
+for all the remaining words in the query which does not fall in another slot category.
+Out of scope slot (O) is a part of slot dictionary that the model is trained on.
+
+Example of model configuration file for training the model could be found at: `NeMo/examples/nlp/intent_slot_classification/conf/intent_slot_classification.yaml <https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/token_classification/conf/token_classification_config.yaml>`__.
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:107]: wrong link

File: docs/source/nlp/joint_intent_slot.rst [hunk 6]
```diff
@@ -3,10 +3,253 @@
...
+
+More details about parameters in the spec file could be found below:
+
++-------------------------------------------+-----------------+----------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
+| **Parameter**                             | **Data Type**   |   **Default**                                                                    | **Description**                                                                                              |
++-------------------------------------------+-----------------+----------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
+| save_to                                   | string          | trained-model.tlt                                                                | The filename of the trained model                                                                            |
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:121]: please remove all tlt related parameters

File: docs/source/nlp/joint_intent_slot.rst [hunk 7]
```diff
@@ -3,10 +3,253 @@
...
+| model.language_model.lm_checkpoint        | string          | null                                                                             | The path to the pre-trained language-model checkpoint                                                        |
++-------------------------------------------+-----------------+----------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
+| model.language_model.config_file          | string          | null                                                                             | The path to the pre-trained language-model config file                                                       |
++-------------------------------------------+-----------------+----------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:150]: Please remove all tokenizer and language_model related params and add a link to the https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/nlp_model.rst#model-nlp

File: docs/source/nlp/joint_intent_slot.rst [hunk 8]
```diff
@@ -3,10 +3,253 @@
...
+| validation_ds.num_workers                 | integer         | 2                                                                                | The number of worker threads for validation                                                                  |
++-------------------------------------------+-----------------+----------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------+
+| validation_ds.batch_size                  | integer         | 32                                                                               | The validation data batch size                                                                               |
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:167]: please remove training/validation_ds params that are common for other models, probably only *.prefix should stay

File: docs/source/nlp/joint_intent_slot.rst [hunk 9]
```diff
@@ -3,10 +3,253 @@
...
+
+Model Evaluation and Inference
+------------------------------
+
+There is no separate script for the evaluation and inference of this model in NeMo, but inside of the example file:
+`examples/nlp/intent_slot_classification/intent_slot_classification.py`
+after the training part is finished you can see the code that evaluates the trained model
+on an evaluation test set and then an example of doing inference using a list of given queries.
+
+For the deployment in the production environment please refer to Jarvis and TLT documentation.
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:255]: @vladgets the links are missing

@ericharper @okuchaiev I don't see that we mention Jarvis and/or TLT in the intro.rst? Any plans on adding this?

**Review by ekmb — CHANGES_REQUESTED** [review]

File: docs/source/nlp/joint_intent_slot.rst (hunk 1 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:45]: Done

Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:45]: Done

File: docs/source/nlp/joint_intent_slot.rst (hunk 3 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:8]: Done

File: docs/source/nlp/joint_intent_slot.rst (hunk 5 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:107]: Corrected

File: docs/source/nlp/joint_intent_slot.rst (hunk 6 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:121]: Corrected

File: docs/source/nlp/joint_intent_slot.rst (hunk 7 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:150]: Corrected

File: docs/source/nlp/joint_intent_slot.rst (hunk 8 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:167]: Corrected

File: docs/source/nlp/joint_intent_slot.rst (hunk 9 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:255]: I do not know the correct links

Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:255]: Ok, I added some links to Jarvis and TLT docs

File: docs/source/nlp/joint_intent_slot.rst (hunk 4 above)
Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst:100]: Done

File: docs/source/nlp/joint_intent_slot.rst [hunk 10]
```diff
@@ -97,14 +117,14 @@ Model Training
 This is a pretrained Bert based model with 2 linear classifier heads on the top of it,
 one for classifying an intent of the query and another for classifying slots for each token of the query.
 This model is trained with the combined loss function on the Intent and Slot classification task on the given dataset.
-The model architecture is based on the paper: https://arxiv.org/pdf/1902.10909.pdf
+The model architecture is based on the paper `BERT for Joint Intent Classification and Slot Filling <https://arxiv.org/pdf/1902.10909.pdf>`__.
```
Comment (ekmb) [inline docs/source/nlp/joint_intent_slot.rst:120]: Could you please add this to the list of references?

Comment (vladgets) [reply docs/source/nlp/joint_intent_slot.rst]: Added

**Review by ekmb — APPROVED** [review]

**Pull request merged**